test/samples/forall_simple.instrumented.sol

```diff
@@ -0,0 +1,355 @@
+                emit AssertionFailed("1: ");
+                assert(false);
+            }
+        }
+        if (_v.__scribble_check_invs_at_end) __scribble_check_state_invariants();
+        __scribble_out_of_contract = _v.__scribble_check_invs_at_end;
+    }
+
+    function _original_ForallSimple_test1(uint[] memory a, uint x) private {}
+
+    function test2(uint[] memory a, uint x) public {
+        vars3 memory _v;
+        _v.__scribble_check_invs_at_end = __scribble_out_of_contract;
+        __scribble_out_of_contract = false;
+        unchecked {
+            _v.old_0 = t;
+        }
+        _original_ForallSimple_test2(a, x);
+        unchecked {
+            _v.forall_2 = true;
+            for (_v.i2 = 0; _v.i2 < a.length; _v.i2++) {
+                _v.forall_2 = a[_v.i2] > _v.old_0;
+                if (!_v.forall_2) break;
+            }
+            if (!(_v.forall_2)) {
+                emit AssertionFailed("2: ");
+                assert(false);
+            }
+        }
+        if (_v.__scribble_check_invs_at_end) __scribble_check_state_invariants();
+        __scribble_out_of_contract = _v.__scribble_check_invs_at_end;
+    }
+
+    function _original_ForallSimple_test2(uint[] memory a, uint x) private {
+        t = x;
+    }
+
+    function test3(uint[] memory a, uint x) public {
+        vars4 memory _v;
+        _v.__scribble_check_invs_at_end = __scribble_out_of_contract;
+        __scribble_out_of_contract = false;
+        unchecked {
+            _v.forall_3 = true;
+            for (_v.i3 = 0; _v.i3 < a.length; _v.i3++) {
+                _v.x1 = a[_v.i3];
+                _v.y = t1;
+                _v.let_2 = _v.x1 > _v.y;
+                _v.let_3 = _v.let_2;
+                _v.forall_3 = _v.let_3;
+                if (!_v.forall_3) break;
+            }
+            _v.old_1 = _v.forall_3;
+        }
+        _original_ForallSimple_test3(a, x);
+        unchecked {
+            if (!(_v.old_1)) {
+                emit AssertionFailed("3: ");
+                assert(false);
+            }
+        }
+        if (_v.__scribble_check_invs_at_end) __scribble_check_state_invariants();
+        __scribble_out_of_contract = _v.__scribble_check_invs_at_end;
+    }
+
+    function _original_ForallSimple_test3(uint[] memory a, uint x) private {
+        t1 = x;
+    }
+
+    function test4(uint[][] memory a, uint x) public {
+        vars5 memory _v;
+        _v.__scribble_check_invs_at_end = __scribble_out_of_contract;
+        __scribble_out_of_contract = false;
+        unchecked {
+            _v.old_2 = t2;
+        }
+        _original_ForallSimple_test4(a, x);
+        unchecked {
+            _v.forall_4 = true;
+            for (_v.i4 = 0; _v.i4 < a.length; _v.i4++) {
+                _v.forall_5 = true;
+                for (_v.j0 = 0; _v.j0 < a[_v.i4].length; _v.j0++) {
+                    _v.forall_5 = a[_v.i4][_v.j0] > _v.old_2;
+                    if (!_v.forall_5) break;
```

Comment:
@cd1m0 I assume in Solidity this breaks only out of the inner loop. Right?

Reply:
@wuestholz It will cause a cascade of `break`s:

```solidity
_v.forall_5 = a[_v.i4][_v.j0] > _v.old_2;
if (!_v.forall_5) break;

// ...

_v.forall_4 = _v.forall_5;
if (!_v.forall_4) break;

// ...

if (!(_v.forall_4)) {
    emit AssertionFailed("4: ");
    assert(false);
}
```

I guess we better state such behavior in the docs (I mean nested `forall`s).

Reply:
👍 Right, that's also what I assumed. I was worried about the semantics of nested breaks since, as far as I remember, some languages automatically break out of the outer loops as well.

Reply:
Solidity `break` breaks only out of the innermost loop.